api/providers/tasks.py

```diff
@@ -0,0 +1,592 @@
+    if not dry_run:
+        try:
+            upload.save()
+        except ValidationError:
+            sentry.log_exception()
+            message = 'Bulk upload preparation failure: failed to create the job'
+            return handle_internal_error(initiator=initiator, provider=provider, message=message, dry_run=dry_run)
+        upload.reload()
+        logger.info('Bulk upload job created: [pk={}, hash={}]'.format(upload.id, upload.payload_hash))
+    else:
+        logger.info('Dry run: insertion did not happen')
+
+    # Create registration rows for the bulk upload job
+    registration_rows = parsing_output.get('registrations', [])
+    if not registration_rows:
+        message = 'Bulk upload preparation failure: missing registration rows'
+        return handle_internal_error(initiator=initiator, provider=provider, message=message, dry_run=dry_run)
+    initial_row_count = len(registration_rows)
+    logger.info('Preparing [{}] registration rows for bulk creation ...'.format(initial_row_count))
+
+    row_hash_set = set()
+    bulk_upload_rows = []
+    draft_error_list = []
+    try:
+        for registration_row in registration_rows:
+            bulk_upload_row = RegistrationBulkUploadRow.create(
+                upload,
+                registration_row.get('csv_raw', ''),
+                registration_row.get('csv_parsed'),
+            )
+            metadata = bulk_upload_row.csv_parsed.get('metadata', {}) or {}
+            row_external_id = metadata.get('External ID', 'N/A')
+            row_title = metadata.get('Title', 'N/A')
+            # Check duplicates with the database
+            if RegistrationBulkUploadRow.objects.filter(row_hash=bulk_upload_row.row_hash).exists():
+                error = 'Duplicate rows - existing row found in the system'
+                exception = RegistrationBulkCreationRowError(upload.id, 'N/A', row_title, row_external_id, error=error)
+                logger.error(exception.long_message)
+                sentry.log_message(exception.long_message)
+                draft_error_list.append(exception.short_message)
+            # Continue to check duplicates within the CSV
+            if bulk_upload_row.row_hash in row_hash_set:
+                error = 'Duplicate rows - CSV contains duplicate rows'
+                exception = RegistrationBulkCreationRowError(upload.id, 'N/A', row_title, row_external_id, error=error)
+                logger.error(exception.long_message)
+                sentry.log_message(exception.long_message)
+                draft_error_list.append(exception.short_message)
+            else:
+                row_hash_set.add(bulk_upload_row.row_hash)
+                bulk_upload_rows.append(bulk_upload_row)
+    except Exception as e:
+        upload.delete()
+        return handle_internal_error(initiator=initiator, provider=provider, message=repr(e), dry_run=dry_run)
+
+    # Cancel the preparation task if duplicates are found in the CSV and/or in DB
```

Comment:
Yes, that was my first thought given your FlowDock example.

However, we want to let user know which rows duplicate or conflict. With that being done preemptively, the bulk_create shouldn't fail. That's way I didn't add the "ignore conflicts" since I want the exception to be caught.

And I will use the `__eq__` and `__hash__` to reduce length code and unnecessary local vars (probably a post-release thing if I don't have time to test in time.)

Reply:
Yeah, graceful error handling and reporting on what aspect of the input caused it to fail is more important than correcting for the bad input